fix(CI): Add a need condition to force tests to start earlier

chouetz opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.

-->
### What does this PR do?
Add a `needs` condition to save time in the overall CI process. Without definition, jobs will wait for completion of all the previous stages, whereas they could nearly all start together.
<img width="773" alt="image" src="https://github.com/DataDog/datadog-agent/assets/24426034/e189d94a-1cac-4f24-8e26-88b0b480f5d4">


<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation
Improve overall CI duration

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes
I've forced the default `new-e2e` template with `deploy_rpm_testing-a6_x64` and `deploy_rpm_testing-a7_x64` as they appear as the last jobs from the `kitchen_deploy` stage. This is maybe fragile, we might want to put all the jobs of the stage to be sure

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

chouetz committed:
fix(CI): Add a need condition to force tests to start earlier

chouetz commented on .gitlab/e2e.yml:
```diff
@@ -69,16 +70,18 @@ k8s-e2e-otlp-dev:
   stage: e2e
   image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/test-infra-definitions/runner$TEST_INFRA_DEFINITIONS_BUILDIMAGES_SUFFIX:$TEST_INFRA_DEFINITIONS_BUILDIMAGES
   tags: ["arch:amd64"]
+  needs:
+    - job: "deploy_rpm_testing-a6_x64" # This deploy is the last one of all the deploys
+      optional: true
+    - job: "deploy_rpm_testing-a7_x64" # In case agent6 is not built
+      optional: true
```
As per PR description, this is maybe fragile, we might want to put all the jobs of the `kitchen_deploy` stage. Jobs having a needs currently depended only on `deploy_windows` and `deploy_deb_a7_x64`

KSerrania replied:
Yes, this is very fragile, it relies on the assumption that RPM builds and deploys are always slower than the rest, which doesn't hold in the general case.

IMO, we could add all `kitchen_deploy` jobs here as a stopgap, but in practice, individual jobs should know what they need, and describe their dependencies accordingly.

kkhor-datadog replied:
If the jobs with dependencies explicitly specified currently depend only on `deploy_windows` and `deploy_deb_a7_x64`, why wouldn't we just specify those same dependencies here? For reference our `new-e2e-process-dev` job uses those same dependencies: https://github.com/DataDog/datadog-agent/blob/9b07d3e8bd6fe91f2fac529db7c954fa1dc6580a/.gitlab/e2e.yml#L281

chouetz replied:
> Member

I was proposing to change this because all the `new-e2e` tests here depend on the 2 same jobs from `kitchen_deploy`, which looks like a bad copy/paste. As Kylian said, we would probably prefer something more explicit, each test should depend on some specific package and/or image build.

kkhor-datadog replied:
For the Windows dependencies at least, they were explicitly added to address flakes in Windows tests: https://github.com/DataDog/datadog-agent/pull/21863

KSerrania commented on .gitlab/e2e.yml:
```diff
@@ -2,11 +2,18 @@
 # e2e stage
 # Contains jobs which runs e2e tests on our Docker images.
 
+.container_needs:
```
I'd make this more explicit, with a name like `.container_e2e_needs` or `.e2e_container_needs`. The reason being that this block _could_ be mistakenly used elsewhere.

KevinFairise2 commented on .gitlab/e2e.yml:
```diff
@@ -2,11 +2,18 @@
   image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-builders/gitlab_agent_deploy:$DATADOG_AGENT_BUILDERS
   tags: ["arch:amd64"]
   dependencies: []
+  extends: .container_needs
```
The `k8s_e2e*` jobs do not refer to jobs run with `new-e2e`. So I am not sure these jobs actually use the images pushed on the `qa` environment

chouetz replied:
the jobs extending `k8s_e2e_template` seem to refer to agent-image and dca-image built during these steps, that's why I put this reference

moth-a reviewed (approved):
Ty, this looks good for the `new-e2e-remote-config` job

chouetz committed:
codereview: improve needs condition and change extension naming

L3n41c commented on .gitlab/e2e.yml:
```diff
@@ -2,11 +2,18 @@
 # e2e stage
 # Contains jobs which runs e2e tests on our Docker images.
 
+.container_e2e_needs:
+  needs:
+    - qa_agent
+    - qa_dca
+    - qa_dogstatsd
+
 .k8s_e2e_template:
   stage: e2e
   image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-builders/gitlab_agent_deploy:$DATADOG_AGENT_BUILDERS
   tags: ["arch:amd64"]
   dependencies: []
+  extends: .container_e2e_needs
```
There is some confusion between the legacy and the new e2e tests here.

### Legacy e2e tests are using

* `datadog/agent-dev:${CI_COMMIT_REF_SLUG}-py3`
* `datadog/cluster-agent-dev:${CI_COMMIT_REF_SLUG}`

images. See [line 37](https://github.com/DataDog/datadog-agent/pull/22817/files#diff-ef82ddf42fac4f3d2cea4532ca305be2ca01345a7f3a471ffb6652f8a5feded5R37) for ex.

Those images are published by the
* `dev_branch_multiarch-a7` https://github.com/DataDog/datadog-agent/blob/0be86ff70fe5748cac248b53f90b4cd8e5eba150/.gitlab/dev_container_deploy/docker_linux.yml#L71
* `dca_dev_branch` https://github.com/DataDog/datadog-agent/blob/0be86ff70fe5748cac248b53f90b4cd8e5eba150/.gitlab/dev_container_deploy/docker_linux.yml#L145
jobs

### New e2e tests are using

* `669783387624.dkr.ecr.us-east-1.amazonaws.com/agent:${CI_PIPELINE_ID}-${CI_COMMIT_SHORT_SHA}`
* `669783387624.dkr.ecr.us-east-1.amazonaws.com/cluster-agent:${CI_PIPELINE_ID}-${CI_COMMIT_SHORT_SHA}`
* `669783387624.dkr.ecr.us-east-1.amazonaws.com/dogstatsd:${CI_PIPELINE_ID}-${CI_COMMIT_SHORT_SHA}`

images.

Those images are indeed published by
* `qa_agent`
* `qa_dca`
* `qa_dogstatsd`
jobs.

## Suggestion:

```suggestion
.container_new_e2e_needs:
  needs:
    - qa_agent
    - qa_dca
    - qa_dogstatsd

.k8s_e2e_template:
  stage: e2e
  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-builders/gitlab_agent_deploy:$DATADOG_AGENT_BUILDERS
  tags: ["arch:amd64"]
  dependencies: []
  needs:
    - dev_branch_multiarch-a7
    - dca_dev_branch
```

chouetz replied:
Added the correct dependencies for the `legacy e2e`

L3n41c commented on .gitlab/e2e.yml:
```diff
@@ -32,9 +39,6 @@ k8s-e2e-cws-dev:
 k8s-e2e-cws-main:
   extends: .k8s_e2e_template
   rules: !reference [.on_main]
-  # needs:
-  #   - dev_master-a6
-  #   - dev_master-a7
   retry: 1
   script:
     - !reference [.k8s-e2e-cws-cspm-init]
```
This job is using
* `datadog/agent-dev:master-py3`
* `datadog/cluster-agent-dev:master`

that are published by:

```suggestion
k8s-e2e-cws-main:
  extends: .k8s_e2e_template
  rules: !reference [.on_main]
  needs:
    - dev_master-a7
    - dca_dev_master
  retry: 1
  script:
    - !reference [.k8s-e2e-cws-cspm-init]
```

See https://github.com/DataDog/datadog-agent/blob/0be86ff70fe5748cac248b53f90b4cd8e5eba150/.gitlab/dev_container_deploy/docker_linux.yml#L120 and https://github.com/DataDog/datadog-agent/blob/0be86ff70fe5748cac248b53f90b4cd8e5eba150/.gitlab/dev_container_deploy/docker_linux.yml#L168

L3n41c reviewed (changes requested):
I left few comments below but there are also issues in pieces that I couldn’t comment.

For ex., the `k8s-e2e-cws-dev` job contains an explicit `needs: []` statement that will override the `needs` clause inherited from the `.k8s_e2e_template` template.

chouetz committed:
codereview: use correct dependency for old e2e fwk. Improve code

chouetz commented:
> I left few comments below but there are also issues in pieces that I couldn’t comment.
> 
> For ex., the `k8s-e2e-cws-dev` job contains an explicit `needs: []` statement that will override the `needs` clause inherited from the `.k8s_e2e_template` template.

Thanks for your review. As discussed, I updated the PR:
- added correct dependency for `k8s-e2e` tests, with templates for `main` and `dev` parts using `variables` for more factorization
- Created 2 `new-e2e-template-needs-*` template to handle needs on `kitchen_deploy` stage or `deploy_containers` one

chouetz commented on .gitlab/e2e.yml:
```diff
@@ -21,59 +21,61 @@
-  extends: .k8s_e2e_template
+  extends: .k8s_e2e_template_dev
   rules: !reference [.on_dev_branch_manual]
-  needs: [] # we can't explicitly define the dependencies because a job cannot depend on other manual jobs.
```
According to [this article](https://about.gitlab.com/blog/2021/05/20/dag-manual-fix/) this should be possible now

chouetz commented on .gitlab/e2e.yml:
```diff
@@ -275,52 +302,42 @@ new-e2e-cws-main:
     TARGETS: ./tests/orchestrator
     TEAM: container-app
 
-  allow_failure: true #TODO: Remove when https://github.com/DataDog/datadog-agent/pull/22113 is merged
```
It was [merged](https://github.com/DataDog/datadog-agent/pull/22113)

chouetz commented:
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=auto

chouetz committed:
Prevent otlp e2e from having inexpected secrets

CelianR reviewed (approved):
Lgtm !

chouetz commented:
/merge
